### DISCUSSION TITLE
Does the p2 downloader respect the offline flag?

### DISCUSSION THREAD

**Discussion by lppedd** [eclipse-tycho/tycho#1801] [Q&A] [answered]
Even though I'm using the `-o` flag on the command line, the p2 log still comes up like the following:
```
[INFO] Downloading from p2: https://download.eclipse.org/nattable/releases/2.0.0/repository/p2.index
[INFO] Downloaded from p2: https://download.eclipse.org/nattable/releases/2.0.0/repository/p2.index (172 bytes at 0 bytes/s)
[INFO] Adding repository https://download.eclipse.org/nattable/releases/2.0.0/repository
```
Are there limitations on the offline capabilities that you know of?

**Comment by laeubi** [comment, marked as answer]
The offline flag is handled at the cache level, so if you are using offline mode and the file is not in the cache you will see an exception, the download message is still printed. One might think about suppress the message in offline mode but I think it does not really harm much.

**Reply by lppedd** [reply]
> it does not really harm much

No no, it doesn't hurt, but it can be misleading, especially because it reports the download speed.

**Reply by laeubi** [reply]
If you like you can open a PR so the message is suppressed in offline mode, it should be rather straight forward ...

**Reply by lppedd** [reply]
@laeubi sounds good. I'll open it soon.

**Reply by lppedd** [reply]
@laeubi ok now I got a better understanding of why the cache doesn't work in my case. I'm using FTP for some of them. I guess there is no way to get FTP working as HTTP (in terms of caching, I see there is an explicit check for "is HTTP"), right?

**Reply by laeubi** [reply]
Currently Tycho has a specialized cache for HTTP but there is no reason we cannot add another one for FTP

https://github.com/eclipse-tycho/tycho/blob/0bff0de2138c414923f00ab6d568c080d21acd2c/p2-maven-plugin/src/main/java/org/eclipse/tycho/p2maven/transport/TychoRepositoryTransport.java#L173

could have another case for `isFTP` and then we could have a similar implementation for FTP, so if you like contributions are welcome as FTP has listing capabilities to query file properties it should even be quite straight forward and one can even have use some tricks compared to http.

For example the very first file requested is the `p2.index` so a smart implementation can (instead of blindly fetch the file first list the directory and store these data to see if the file has changed and as P2 likely will request other files from this directory as well.

I think the most important part will be to keep the connections alive as for FTP the connection setup is even more expensive than for http.
